opencog/attentionbank/AttentionSCM.cc

```diff
@@ -0,0 +1,214 @@
+	define_scheme_primitive("cog-inc-vlti!", &AttentionSCM::inc_vlti, this, "attention-bank");
+	define_scheme_primitive("cog-dec-vlti!", &AttentionSCM::dec_vlti, this, "attention-bank");
+
+	// Attentional-focus-related stuff
+	define_scheme_primitive("cog-update-af", &AttentionSCM::update_af, this, "attention-bank");
+	define_scheme_primitive("cog-af-size", &AttentionSCM::af_size, this, "attention-bank");
+	define_scheme_primitive("cog-set-af-size!", &AttentionSCM::set_af_size, this, "attention-bank");
+	define_scheme_primitive("cog-stimulate", &AttentionSCM::stimulate, this, "attention-bank");
+}
+
+AttentionSCM::~AttentionSCM()
+{
+}
+
+AttentionValuePtr AttentionSCM::get_av(const Handle& h)
+{
+	return opencog::get_av(h);
+}
+
+Handle AttentionSCM::set_av(const Handle& h, const AttentionValuePtr& av)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-set-av!");
+	attentionbank(atomspace).change_av(h, av);
+	return h;
+}
+
+Handle AttentionSCM::inc_vlti(const Handle& h)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-inc-vlti!");
+	attentionbank(atomspace).inc_vlti(h);
+	return h;
+}
+
+Handle AttentionSCM::dec_vlti(const Handle& h)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-dec-vlti!");
+	attentionbank(atomspace).dec_vlti(h);
+	return h;
+}
+
+/**
+ *   Return AttentionalFocus Size
+ **/
+int AttentionSCM::af_size(void)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-af-size");
+	return attentionbank(atomspace).get_af_size();
+}
+
+/**
+ * Set AttentionalFocus Size
+ */
+int AttentionSCM::set_af_size (int ssize)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-set-af-size!");
+
+	attentionbank(atomspace).set_af_size(ssize);
+	return attentionbank(atomspace).get_af_size();
+}
+
+/**
+ * Return the list of top n atoms in the AttentionalFocus or
+ * return all atoms in the AF if n is unspecified or is larger
+ * than the AF size.
+ */
+Handle AttentionSCM::update_af(int n)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-af");
+
+	Handle af_anchor = atomspace->add_node(ANCHOR_NODE,
+	                               "*-attentional-focus-boundary-*");
+
+	// Get the atoms that were previously in attention focus
+	IncomingSet paf(af_anchor->getIncomingSetByType(MEMBER_LINK));
+	HandleSet prev_af;
+	for (const LinkPtr& lp: paf)
+	{
+		Handle h(lp->getOutgoingAtom(0));
+		if (h != af_anchor)
+			prev_af.insert(h);
+	}
+
+	HandleSeq attentionalFocus;
+	attentionbank(atomspace).get_handle_set_in_attentional_focus(back_inserter(attentionalFocus));
+	size_t isz = attentionalFocus.size();
+
+	size_t N = isz;
+	if (0 < n) N = n;
+	if( N > isz)  N = isz;
+
+	/* Add all the atoms in the current attentional focus that are not
+	 * already there. */
+	HandleSet curr_af;
+	for (size_t i = isz - N; i < isz; i++)
+	{
+		Handle hi = attentionalFocus[i];
+		auto gone = prev_af.find(hi);
+		if (prev_af.end() == gone)
+		{
+			atomspace->add_link(MEMBER_LINK, hi, af_anchor);
```

Comment:
This could create a nested list of `MemberLinks` if `prev_af` has `MemberLink`s used to represent the set of atoms in the attentional-focus(af). The logic = the representative sets are likely to be in the af because they are connected to atoms that are in the af, and `sti` will diffuse to them as a result of which they are likely to enter the af.